File: Bidding/ClientApp/src/app/auctions/services/auctions.service.ts
// angular
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';

// 3rd lib
import { catchError } from 'rxjs/operators';
import { Observable } from 'rxjs';

// internal
import { AuctionListResponseModel } from '../models/list/auction-list-response.model';
import { AuctionListRequestModel } from '../models/list/auction-list-request.model';
import { AuctionDetailsModel } from '../models/details/auction-details.model';
import { ExceptionsService } from '../../core/services/exceptions/exceptions.service';
import { AuctionEditRequestModel } from '../models/edit/auction-edit-request.model';
import { AuctionFilterModel } from '../models/filters/auction-filter.model';
import { AuctionFormatModel } from '../models/add/auction-format.model';
import { AuctionCreatorModel } from '../models/add/auction-creator.model';
import { AuctionStatusModel } from '../models/add/auction-status.model';
import { AuctionDeleteRequest } from '../models/delete/auction-delete-request.model';
import { CategoriesWithTypesModel } from '../models/add/categories-with-types.model';
import { AuctionEditDetailsResponseModel } from '../models/edit/auction-edit-details-response.model';


@Injectable({
  providedIn: 'root'
})
export class AuctionsService {
  constructor(
    private httpService: HttpClient,
    private exceptionService: ExceptionsService
  ) { }

  getAuctions$(request: AuctionListRequestModel): Observable<AuctionListResponseModel> {
    const url = '/api/auctions/list';

    let params = new HttpParams({
      fromObject: {
        sortByColumn: request.sortByColumn.toString(),
        sortingDirection: request.sortingDirection.toString(),
        offsetEnd: request.offsetEnd.toString(),
        offsetStart: request.offsetStart.toString(),
        searchValue: request.searchValue === undefined ? '' : request.searchValue.toString(),
        currentPage: request.currentPage.toString()
      }
    });

    if (request.topCategoryIds !== undefined) {
      for (const id of request.topCategoryIds) {
        params = params.append('topCategoryIds', id.toString());
      }
    }

    if (request.typeIds !== undefined) {
      for (const id of request.typeIds) {
        params = params.append('typeIds', id.toString());
      }
    }

    return this.httpService.get<AuctionListResponseModel>(url, { params })
      .pipe(catchError(this.exceptionService.errorHandler));
  }

  /** Loads filters for auction list */
  getFilters$(): Observable<AuctionFilterModel> {
    const url = '/api/auctions/filters';

    return this.httpService.get<AuctionFilterModel>(url)
      .pipe(catchError(this.exceptionService.errorHandler));
  }

  /** Fetch all top-categories with sub-categories / types for auction add wizard */
  categoriesWithTypes$(): Observable<CategoriesWithTypesModel> {
    const url = '/api/auctions/CategoriesWithTypes';

    return this.httpService.get<CategoriesWithTypesModel>(url)
      .pipe(catchError(this.exceptionService.errorHandler));
  }

  getAuctionCreators$(): Observable<AuctionCreatorModel[]> {
    const url = '/api/auctions/creators';

    return this.httpService.get<AuctionCreatorModel[]>(url)
      .pipe(catchError(this.exceptionService.errorHandler));
  }

  getAuctionFormats$(): Observable<AuctionFormatModel[]> {
    const url = '/api/auctions/formats';

    return this.httpService.get<AuctionFormatModel[]>(url)
      .pipe(catchError(this.exceptionService.errorHandler));
  }

  getAuctionStatuses$(): Observable<AuctionStatusModel[]> {
    const url = '/api/auctions/statuses';

    return this.httpService.get<AuctionStatusModel[]>(url)
      .pipe(catchError(this.exceptionService.errorHandler));
  }

  getAuctionDetails$(auctionId: number): Observable<AuctionDetailsModel> {
    const url = `api/auctions/details?auctionId=${auctionId}`;

    return this.httpService.get<AuctionDetailsModel>(url)
      .pipe(catchError(this.exceptionService.errorHandler));
  }

  /**
   * Adds a new auction, it can be item, vehicle or property - auction
   * @param request Global auction add request
   */
  addAuction$(request: Auctions.AddAuctionRequestModel): Observable<boolean> {
    const url = '/api/auctions/create';

    return this.httpService.post<boolean>(url, request)
      .pipe(catchError(this.exceptionService.errorHandler));
  }

  getAuctionEditDetails$(auctionId: number): Observable<AuctionEditDetailsResponseModel> {
    const url = `api/auctions/editDetails?auctionId=${auctionId}`;

    return this.httpService.get<AuctionEditDetailsResponseModel>(url)
      .pipe(catchError(this.exceptionService.errorHandler));
  }

  editAuction$(request: AuctionEditRequestModel): Observable<boolean> {
    const url = '/api/auctions/edit';

    return this.httpService.put<boolean>(url, request)
      .pipe(catchError(this.exceptionService.errorHandler));
  }

  deleteAuction$(request: AuctionDeleteRequest): Observable<boolean> {
    const url = '/api/auctions/delete';
    const options = {
      headers: {},
      body: request
    };

    return this.httpService.delete<boolean>(url, options)
      .pipe(catchError(this.exceptionService.errorHandler));
  }
}

File: Bidding/Services/Auctions/AuctionsService.cs
using Bidding.Models.DatabaseModels.Auctions;
using Bidding.Models.ViewModels.Auctions.Add;
using Bidding.Models.ViewModels.Auctions.Delete;
using Bidding.Models.ViewModels.Auctions.Details;
using Bidding.Models.ViewModels.Auctions.Edit;
using Bidding.Models.ViewModels.Auctions.List;
using Bidding.Models.ViewModels.Auctions.Shared;
using Bidding.Models.ViewModels.Filters;
using Bidding.Repositories.Auctions;
using Bidding.Services.Shared;
using Bidding.Services.Shared.Permissions;
using Bidding.Shared.Constants;
using Bidding.Shared.ErrorHandling.Errors;
using Bidding.Shared.ErrorHandling.Validators;
using Bidding.Shared.Exceptions;
using Bidding.Shared.Pagination;
using Bidding.Shared.Utility.Validation.Comparers;
using FluentValidation;
using FluentValidation.Results;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Threading.Tasks;

namespace Bidding.Services.Auctions
{
    public class AuctionsService
    {
        private readonly PermissionService m_permissionService;
        private readonly AuctionsRepository m_auctionsRepository;
        private readonly FileUploaderService m_fileUploaderService;

        public AuctionsService(AuctionsRepository auctionRepository, PermissionService permissionService, FileUploaderService fileUploaderService)
        {
            m_permissionService = permissionService ?? throw new ArgumentNullException(nameof(permissionService));
            m_auctionsRepository = auctionRepository ?? throw new ArgumentNullException(nameof(auctionRepository));
            m_fileUploaderService = fileUploaderService ?? throw new ArgumentNullException(nameof(fileUploaderService));
        }

        public AuctionListResponseModel GetAuctions(AuctionListRequestModel request)
        {
            ValidateAuctionList(request);

            // note: kke: Always set searchValue to be null here, 
            // because the user needs to be signed-in to use search functionality!
            request.SearchValue = null;

            (int startFrom, int endAt) = Pagination.GetStartAndEnd(request);

            AuctionListResponseModel auctionsResponse = new AuctionListResponseModel()
            {
                Auctions = m_auctionsRepository.GetAuctions(request, startFrom, endAt).ToList(),
                ItemCount = m_auctionsRepository.TotalAuctionCount().Count()
            };

            Pagination.PaginateResponse(ref auctionsResponse, TableItem.DEFAULT_SIZE, request.CurrentPage);

            return auctionsResponse;
        }

        public AuctionListResponseModel GetAuctionsWithSearch(AuctionListRequestModel request)
        {
            ValidateAuctionList(request);

            // todo: kke: validate request.SearchValue to be valid value!

            (int startFrom, int endAt) = Pagination.GetStartAndEnd(request);

            AuctionListResponseModel auctionsResponse = new AuctionListResponseModel()
            {
                Auctions = m_auctionsRepository.GetAuctions(request, startFrom, endAt).ToList(),
                ItemCount = m_auctionsRepository.TotalAuctionCount().Count()
            };

            Pagination.PaginateResponse(ref auctionsResponse, TableItem.DEFAULT_SIZE, request.CurrentPage);

            return auctionsResponse;
        }

        /// <summary>
        /// Loads all filters, top category filter and sub-category filter
        /// </summary>
        /// <returns></returns>
        public AuctionFilterModel Filters()
        {
            return new AuctionFilterModel()
            {
                TopCategories = m_auctionsRepository.LoadActiveTopCategoriesWithCount().ToList(),
                SubCategories = m_auctionsRepository.LoadActiveSubCategoriesWithCount().ToList()
            };
        }

        public CategoriesWithTypesModel CategoriesWithTypes()
        {
            m_permissionService.IsLoggedInUserActive();

            return new CategoriesWithTypesModel()
            {
                TopCategories = m_auctionsRepository.LoadTopCategories().ToList(),
                SubCategories = m_auctionsRepository.LoadSubCategories().ToList()
            };
        }

        public AuctionCreatorModel Creators()
        {
            m_permissionService.IsLoggedInUserActive();

            return new AuctionCreatorModel()
            {
                Creators = m_auctionsRepository.Creators().ToList()
            };
        }

        public AuctionFormatModel Formats()
        {
            m_permissionService.IsLoggedInUserActive();

            return new AuctionFormatModel()
            {
                Formats = m_auctionsRepository.Formats().ToList()
            };
        }

        public AuctionFormatModel CreateVehicleDetails()
        {
            m_permissionService.IsLoggedInUserActive();

            return new AuctionFormatModel()
            {
                Formats = new List<AuctionFormatItemModel>() // m_auctionsRepository.CreateVehicleDetails()
            };
        }

        public AuctionStatusModel Statuses()
        {
            m_permissionService.IsLoggedInUserActive();

            return new AuctionStatusModel()
            {
                Statuses = m_auctionsRepository.Statuses().ToList()
            };
        }

        public async Task<AuctionDetailsResponseModel> DetailsAsync(AuctionDetailsRequestModel request)
        {
            if (request.IsNotSpecified()) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }
            if (request.AuctionId.IsNotSpecified()) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.IncorrectAuction); }

            // var xxx = await m_fileUploaderService.GetCloudBlobContainer2();

            return await m_auctionsRepository.DetailsAsync(request);
        }

        public AuctionEditDetailsResponseModel EditDetails(int auctionId)
        {
            if (auctionId.IsNotSpecified()) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }

            m_permissionService.IsLoggedInUserActive();

            return m_auctionsRepository.EditDetails(auctionId);
        }

        public bool UpdateAuctionDetails(AuctionEditRequestModel request)
        {
            ValidateAuctionUpdate(request);

            int? loggedInUserId = m_permissionService.GetUserIdFromClaimsPrincipal();

            return m_auctionsRepository.UpdateAuctionDetails(request, loggedInUserId.Value);
        }

        public bool Delete(AuctionDeleteRequestModel request)
        {
            ValidateAuctionDelete(request);

            int? loggedInUserId = m_permissionService.GetUserIdFromClaimsPrincipal();

            return m_auctionsRepository.Delete(request, loggedInUserId.Value);
        }

        public bool Create(AddAuctionRequestModel request)
        {
            if (request.IsNotSpecified()) throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation);
            if (request.AboutAuction.AuctionTopCategoryId.IsNotSpecified()) throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation);

            m_permissionService.IsLoggedInUserActive();

            bool status = false;

            if (request.AboutAuction.AuctionTopCategoryId == Categories.ITEM_CATEGORY)
            {
                status = CreateItemAuction(request);
            }
            else if (request.AboutAuction.AuctionTopCategoryId == Categories.VEHICLE_CATEGORY)
            {
                status = CreateVehicleAuction(request);
            }
            else if (request.AboutAuction.AuctionTopCategoryId == Categories.PROPERTY_CATEGORY)
            {
                status = CreatePropertyAuction(request);
            }
            else
            {
                throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation);
            }

            return status;
        }

        /// <summary>
        /// Validate auction list permissions and input values
        /// </summary>
        /// <param name="request"></param>
        private void ValidateAuctionList(AuctionListRequestModel request)
        {
            if (request.IsNotSpecified()) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }
            if (request.SortingDirection.IsNotSpecified()) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }
            if (request.OffsetStart < 0) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }
            if (request.OffsetEnd.IsNotSpecified()) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }
            if (request.CurrentPage < 0) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }
            if (request.SortingDirection != "asc" && request.SortingDirection != "desc") { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }

            // todo: kke: validate sort by columns list!
            // todo: kke: validate if smth else is not missing here!
            if (request.SortByColumn.IsNotSpecified()) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }
            List<string> allowedSortByColumns = new List<string> { "AuctionName", "AuctionStartingPrice", "AuctionStartDate", "AuctionEndDate" };
            if (allowedSortByColumns.Contains(request.SortByColumn) == false) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }
        }

        private bool CreateItemAuction(AddAuctionRequestModel request)
        {
            // ValidateAuctionItemCreate(request);

            int? loggedInUserId = m_permissionService.GetUserIdFromClaimsPrincipal();

            return m_auctionsRepository.CreateItemAuction(request, loggedInUserId.Value);
        }

        private bool CreatePropertyAuction(AddAuctionRequestModel request)
        {
            // ValidateAuctionItemCreate(request);

            int? loggedInUserId = m_permissionService.GetUserIdFromClaimsPrincipal();

            return m_auctionsRepository.CreatePropertyAuction(request, loggedInUserId.Value);
        }

        private bool CreateVehicleAuction(AddAuctionRequestModel request)
        {
            // ValidateAuctionItemCreate(request);

            int? loggedInUserId = m_permissionService.GetUserIdFromClaimsPrincipal();

            return m_auctionsRepository.CreateVehicleAuction(request, loggedInUserId.Value);
        }

        private void ValidateAuctionItemCreate(string request)
        {
            // if (request.IsNotSpecified()) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }
            // if (request.AuctionName.IsNotSpecified()) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }
            //if (request.AuctionTopCategoryIds.IsNotSpecified()) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }
            //if (request.AuctionSubCategoryIds.IsNotSpecified()) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }
            //if (request.AuctionStartingPrice.IsNotSpecified()) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }
            //if (request.AuctionStartDate.IsNotSpecified()) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }
            //if (request.AuctionApplyTillDate.IsNotSpecified()) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }
            //if (request.AuctionEndDate.IsNotSpecified()) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }
            //if (request.AuctionCreatorId.IsNotSpecified()) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }
            //if (request.AuctionFormatId.IsNotSpecified()) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }

            m_permissionService.IsLoggedInUserActive();

            // todo: kke: save all dates to be UTC Format
            // todo: kke: validate all required fields to be specified + max lenghts / min leghts and regular expressions

            // setup auction object for the validation
            //Auction auction = new Auction()
            //{
            //    Name = request.AuctionName,
            //    StartingPrice = request.AuctionStartingPrice,
            //    StartDate = request.AuctionStartDate
            //};

            //ValidateAuction(auction);
        }

        private void ValidateAuction(Auction auction)
        {
            AuctionValidator validator = new AuctionValidator();
            ValidationResult results = validator.Validate(auction);

            bool success = results.IsValid;
            IList<ValidationFailure> failures = results.Errors;

            validator.ValidateAndThrow(auction);

            if (results.IsValid == false)
            {
                foreach (ValidationFailure failure in results.Errors)
                {
                    throw new WebApiException(HttpStatusCode.BadRequest, failure.ErrorMessage);
                }
            }
        }

        /// <summary>
        /// Validate auction update permissions
        /// </summary>
        /// <param name="request"></param>
        private void ValidateAuctionUpdate(AuctionEditRequestModel request)
        {
            if (request.IsNotSpecified()) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }
            if (request.AuctionId.IsNotSpecified()) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }

            // todo: kke: add missing validation checks!

            m_permissionService.IsLoggedInUserActive();
        }

        private void ValidateAuctionDelete(AuctionDeleteRequestModel request)
        {
            if (request.IsNotSpecified()) { throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation); }
            if (request.AuctionIds.IsNotSpecified()) throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation);
            if (request.AuctionIds.Any(s => s.IsNotSpecified())) throw new WebApiException(HttpStatusCode.BadRequest, AuctionErrorMessages.MissingAuctionsInformation);

            m_permissionService.IsLoggedInUserActive();
        }
    }
}
